ST_resgroups.c

```diff
@@ -185,124 +181,124 @@ void rgroup_PartResources(void)
+	#define xF_DELTA (20*F_DELTA)
+	#define xD_DELTA (20*D_DELTA)
+	#define ZERO(x) \
+    ( (sizeof(x) == sizeof(float)) \
+  ? ((x)>-xF_DELTA && (x)<xF_DELTA) \
+  : ((x)>-xD_DELTA && (x)<xD_DELTA) )
+    IntU i, num_est; //number of individuals that will establish this year
+    RealF viable_seeds, viable_seeds_reduced; //viable seed and viable seed excluding this year's added seed, respectively
+    float alpha, beta, var; //parameters needed to calculation of num_est from beta distribution
+    GroupType *g;
+    SpeciesType *s;
+
+    g = RGroup[rg];
+    assert(g->max_age == 1);
+
+    s = Species[sp];
+    viable_seeds_reduced = 0.0;
+
+    /*Increment viable years for seeds and add seeds to the seedbank*/
+    _add_annual_seedprod(sp, lastyear_relsize);
+
+    /*Implement decay process and then get viable seeds from seed bank*/
+    viable_seeds = (g->regen_ok) ? _get_annual_maxestab(sp) : 0;
+
+    /*Create a beta random number draw, where mean =s->seedling_estab_prob
+     variance = s->var. 
+     Calculate alpha and beta based on the mean and variance values by species.in*/
+     alpha = (pow (s->seedling_estab_prob, 2) - pow (s->seedling_estab_prob, 3)
+             - s->seedling_estab_prob * s->var) / s->var;
+     beta = (s->seedling_estab_prob - pow (s->seedling_estab_prob, 3) 
+             - s->var + s->var * s->seedling_estab_prob)/ s->var;
+     
+     /*Create beta random number*/
+     var = RandBeta(alpha, beta);
+     
+     /*Determine number of seedlings to add*/
+     num_est = viable_seeds * var;
+     	//printf("Species name=%s , num_est   =%u \n",s->name,  num_est);
+     
+     /*if the beta random number * viable seeds is larger than max_seed_estab, 
+     max_seed_estab value is used instead*/
+     if (num_est > s->max_seed_estab) {num_est = s->max_seed_estab;} 
+          //printf("Species name=%s , num_est   =%u \n",s->name,  num_est);
+    
+    /*Remove seedlings added from the seed bank*/
+    /*Calculate the viable seedbank excluding seeds added this year*/
+    viable_seeds_reduced= (IntU) viable_seeds - s->seedprod[0];
+    	//printf("Species name=%s , viable_seeds_reduced= %d \n", s->name, viable_seeds_reduced);
+            
+    /*Multiple the proportion of seeds in each viable year array by the total 
+    number of seeds that germinated as seedlings and subtract those seeds from 
+    the relevant seedprod array.*/
+       for (i = s->viable_yrs; i > 0; i--)
+       {
+        //printf("Species name=%s , old calculated value s->seedprod[%hu]= %d \n", s->name, i, s->seedprod[i]);  
+        s->seedprod[i] =  s->seedprod[i] -  round(num_est * s->seedprod[i] / viable_seeds_reduced); 
+        //printf("Species name=%s , so new calculated value s->seedprod[%hu]= %d \n", s->name, i, s->seedprod[i]);
+       }   
+        return num_est;
 }
 
-static RealF _get_annual_maxestab(SppIndex sp)
-{
-	/*======================================================*/
+static RealF _get_annual_maxestab(SppIndex sp) {
```

Comment:
Why do we do the "decay process" in this function the intent of which is to "get max estab seeds"?

Why not in the function `_add_annual_seedprod` which does already the seed aging process? We could save an entire loop over `s->viable_yrs` if we did the `/ pow(i, s->exp_decay)` in line 286.

Reply:
Good idea. Extra loop removed and the decay process is now incorporated in _add_annual_seedprod. Commit https://github.com/DrylandEcology/STEPWAT2/commit/ab42898998759b0c8c01156263e35314cc459df4